backend/battle/tests/test_view.py

```diff
@@ -74,3 +81,178 @@ def test_verify_if_not_returns_winner_missing_pokemons_teams(self):
+        battle = Battle.objects.filter(creator=self.creator, opponent=self.opponent)
+        self.assertTrue(battle)
+
+    def test_if_returns_error_when_creator_and_opponent_are_the_same_user(self):
+        battle_data = {
+            "creator": self.creator.id,
+            "opponent": self.creator.email,
+        }
+        self.client.login(username=self.creator.email, password='admin')
+        response = self.client.post(reverse('battle'), battle_data)
+        self.assertEqual(
+            response.context_data['form'].errors['__all__'][0],
+            "ERROR: You can't challenge yourself.")
+
+    def test_if_returns_error_when_opponent_doesnt_exist(self):
+        battle_data = {
+            "creator": self.creator.id,
+            "opponent": 100,
+        }
+        self.client.login(username=self.creator.email, password='admin')
+        with self.assertRaises(KeyError):
+            self.client.post(reverse('battle'), battle_data)
+
+    @patch("battle.battles.email.send_templated_mail")
+    def test_if_battle_invitation_email_is_sent(self, mock_templated_mail):
+        battle_data = {
+            "creator": self.creator.id,
+            "opponent": self.opponent.email,
+        }
+
+        self.client.login(username=self.creator.email, password='admin')
+
+        self.client.post(reverse("battle"), battle_data)
+
+        battle = Battle.objects.filter(
+            creator=self.creator, opponent=self.opponent)
+
+        self.assertTrue(battle)
+
+        mock_templated_mail.assert_called_with(
+            template_name="invite_challenge",
+            from_email=settings.FROM_EMAIL,
+            recipient_list=[self.opponent.email],
+            context={
+                'creator': self.creator.email,
+            },
+        )
+
+    @patch('battle.tasks.run_battle')
+    def test_if_run_battle_task(self, result_battle_mock):
+        battle = Battle.objects.create(creator=self.creator, opponent=self.opponent)
+
+        team_creator = baker.make('battle.Team', battle=battle, trainer=self.creator)
+        team_opponent = baker.make('battle.Team', battle=battle, trainer=self.opponent)
+        pokemon_1, pokemon_2, pokemon_3 = baker.make(
+            "pokemon.Pokemon",
+            attack=60, defense=45, hp=50, _quantity=3)
+
+        pokemons = [pokemon_1, pokemon_2, pokemon_3]
+
+        for count, pokemon in enumerate(pokemons):
+            PokemonTeam.objects.create(
+                team=team_creator,
+                pokemon=pokemon,
+                order=count)
+
+        for count, pokemon in enumerate(pokemons):
+            PokemonTeam.objects.create(
+                team=team_opponent,
+                pokemon=pokemon,
+                order=count)
+
+        result_battle_mock.return_value = team_creator
+        run_battle_and_send_result_email.apply([battle.id])
+        result_battle_mock.assert_called_with(battle)
+
+
+class IntegrationPokeapiTest(TestCase):
```

Comment:
Nice, this is the way to do mocks. Now you need to mock `get_pokemon_from_api` on where it's used on the system.

Reply:
Great!